core/drivers/stm32mp15_huk.c

```diff
@@ -0,0 +1,157 @@
+		return TEE_ERROR_GENERIC;
+
+	if (stm32_bsec_set_sw_lock(UID1_OTP) ||
+	    stm32_bsec_shadow_read_otp(q++, UID1_OTP))
+		return TEE_ERROR_GENERIC;
+
+	if (stm32_bsec_set_sw_lock(UID2_OTP) ||
+	    stm32_bsec_shadow_read_otp(q++, UID2_OTP))
+		return TEE_ERROR_GENERIC;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result stm32mp15_read_otp(uint32_t otp, uint32_t *key, bool *locked)
+{
+	bool tmp = true;
+
+	if (!stm32mp_is_closed_device()) {
+		/*
+		 * When the device is not closed, the shadow memory for these
+		 * words might not be locked: check and report them
+		 */
+		if (stm32_bsec_read_permanent_lock(otp, &tmp))
+			return TEE_ERROR_GENERIC;
+
+		if (tmp && stm32_bsec_read_sw_lock(otp, &tmp))
+			return TEE_ERROR_GENERIC;
+	}
+
+	if (stm32_bsec_shadow_read_otp(key, otp))
+		return TEE_ERROR_GENERIC;
+
+	*locked = *locked && tmp;
+
+	return TEE_SUCCESS;
+}
+
+/*
+ *  AES-GCM: nonce must be unique per message and key.
+ *
+ *  This function always uses the same key - once its locked - with the same
+ *  unique message hence the nonce can be any constant.
+ */
+static TEE_Result aes_gcm_encrypt_uid(uint8_t *key, size_t key_len,
+				      uint8_t *out, size_t *out_len)
+{
+	TEE_Result ret = TEE_ERROR_GENERIC;
+	const uint8_t nonce[12] = { 0x55 };
+	uint32_t uid[4] = { 0 };
+	uint8_t tag[16] = { 0 };
+	size_t nonce_len = sizeof(nonce);
+	size_t tag_len = sizeof(tag);
+	size_t uid_len = sizeof(uid);
+	void *ctx = NULL;
+
+	ret = stm32mp15_read_uid(uid);
+	if (ret)
+		goto err;
+
+	ret = crypto_authenc_alloc_ctx(&ctx, TEE_ALG_AES_GCM);
+	if (ret)
+		goto err;
+
+	ret = crypto_authenc_init(ctx, TEE_MODE_ENCRYPT, key, key_len, nonce,
+				  nonce_len, TEE_AES_BLOCK_SIZE, 0, uid_len);
+	if (ret)
+		goto auth_err;
+
+	ret = crypto_authenc_enc_final(ctx, (uint8_t *)uid, sizeof(uid),
+				       out, out_len, tag, &tag_len);
+	if (ret)
+		goto auth_err;
+
+	crypto_authenc_final(ctx);
+auth_err:
+	crypto_authenc_free_ctx(ctx);
+err:
+	if (ret)
+		memzero_explicit(out, *out_len);
+
+	return ret;
+}
+
+TEE_Result tee_otp_get_hw_unique_key(struct tee_hw_unique_key *hwkey)
+{
+	uint32_t otp_key[4] = { 0 };
+	size_t len = sizeof(otp_key);
+	TEE_Result ret = TEE_SUCCESS;
+	uint32_t *key = otp_key;
+	bool lock = true;
+
+	static_assert(CFG_STM32MP15_HUK_BSEC_KEY_0 < STM32MP1_OTP_MAX_ID);
+	static_assert(CFG_STM32MP15_HUK_BSEC_KEY_1 < STM32MP1_OTP_MAX_ID);
+	static_assert(CFG_STM32MP15_HUK_BSEC_KEY_2 < STM32MP1_OTP_MAX_ID);
+	static_assert(CFG_STM32MP15_HUK_BSEC_KEY_3 < STM32MP1_OTP_MAX_ID);
+
+	ret = stm32mp15_read_otp(CFG_STM32MP15_HUK_BSEC_KEY_0, key++, &lock);
+	if (ret)
+		goto out;
+
+	ret = stm32mp15_read_otp(CFG_STM32MP15_HUK_BSEC_KEY_1, key++, &lock);
+	if (ret)
+		goto out;
+
+	ret = stm32mp15_read_otp(CFG_STM32MP15_HUK_BSEC_KEY_2, key++, &lock);
+	if (ret)
+		goto out;
+
+	ret = stm32mp15_read_otp(CFG_STM32MP15_HUK_BSEC_KEY_3, key++, &lock);
+	if (ret)
+		goto out;
+
+	ret = aes_gcm_encrypt_uid((uint8_t *)otp_key, len, hwkey->data, &len);
```

Comment:
LGTM.
As per discussion, I will create a P-R to condition the uid encryption upon `CFG_STM32MP15_HUK_BSEC_DERIVE_UID?=n` so that when enabled, the default HUK used is the one programmed in the fuses. @ldts, is that Ok with you?

Reply:
yep that seems like a good plan. Do you wish to maintain this driver?

Reply:
@etienne-lms ^^

Reply:
Ok for me.

Reply:
ok so when you send your PR maybe you can add yourself to the maintainer file for this one? thanks for this too.